components/core/src/metrics.rs

```diff
@@ -0,0 +1,221 @@
+}
+
+// Supported metrics
+#[derive(Debug, Clone)]
+pub enum Gauge {
+    PackageCount,
+}
+
+// Helper types
+#[derive(Debug, Clone, Copy)]
+enum MetricType {
+    Counter,
+    Gauge,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum MetricOperation {
+    Increment,
+    Decrement,
+    SetValue,
+}
+
+type MetricId = &'static str;
+type MetricValue = f64;
+type MetricTuple = (MetricType, MetricOperation, MetricId, Option<MetricValue>);
+
+trait Metric {
+    fn id(&self) -> &'static str;
+}
+
+// One-time initialization
+static mut SENDER: *const Sender<MetricTuple> = 0 as *const Sender<MetricTuple>;
+
+static INIT: Once = ONCE_INIT;
+
+fn sender() -> Sender<MetricTuple> {
+    unsafe {
+        INIT.call_once(|| {
+            SENDER = Box::into_raw(Box::new(init()));
+        });
+        (*SENDER).clone()
+    }
+}
+
+// init creates a worker thread ready to receive and process metric events,
+// and returns a channel for use by metric senders
+fn init() -> Sender<MetricTuple> {
+    let (tx, rx) = channel::<MetricTuple>();
+    let (rztx, rzrx) = sync_channel(0); // rendezvous channel
+
+    thread::Builder::new()
+        .name("metrics".to_string())
+        .spawn(move || receive(rztx, rx))
+        .expect("couldn't start metrics thread");
+
+    match rzrx.recv() {
+        Ok(()) => tx,
+        Err(e) => panic!("metrics thread startup error, err={}", e),
+    }
+}
+
+// receive runs in a separate thread and processes all metrics events
+fn receive(rz: SyncSender<()>, rx: Receiver<MetricTuple>) {
+    let mut client = statsd_client();
+    rz.send(()).unwrap(); // Blocks until the matching receive is called
+
+    loop {
+        let (mtyp, mop, mid, mval): MetricTuple = rx.recv().unwrap();
+        debug!("Received metrics tuple: {:?}", (mtyp, mop, mid, mval));
+
+        match client {
+            Some(ref mut cli) => {
+                match mtyp {
+                    MetricType::Counter => {
+                        match mop {
+                            MetricOperation::Increment => cli.incr(mid),
+                            MetricOperation::Decrement => cli.decr(mid),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                    MetricType::Gauge => {
+                        match mop {
+                            MetricOperation::SetValue => cli.gauge(mid, mval.unwrap()),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                }
+            }
+            None => (),
+        }
+    }
+}
+
+fn statsd_client() -> Option<Client> {
+    match env::var(STATS_ENV) {
+        Ok(addr) => Some(Client::new(&addr, APP_NAME).unwrap()),
+        Err(_) => None,
+    }
+}
+
+impl Counter {
+    pub fn increment(&self) {
+        sender()
+            .send((MetricType::Counter, MetricOperation::Increment, &self.id(), None))
+            .unwrap();
```

Comment:
Is there a reason that `unwrap()` is used in the client side API? I believe that if this call returns `Err` (such as if the metrics thread crashes and therefore hangs up the receiver side), this would cause the client/calling side to also panic and unwind its thread. If we're okay ignoring failures, it might be better to explicitly squelch any `Err` rather than crashing out the thread.